File: SmartReceipts/Reports/Generators/TripCSVGenerator.m
//
//  TripCSVGenerator.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 25/04/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "TripCSVGenerator.h"
#import "ReportCSVTable.h"
#import "WBPreferences.h"
#import "NSMutableString+Extensions.h"

@implementation TripCSVGenerator

- (BOOL)generateToPath:(NSString *)outputPath {
    NSString *content = [self generateContent];
    
    NSError *writeError = nil;
    BOOL writeSuccess = [content writeToFile:outputPath atomically:YES encoding:NSUTF8StringEncoding error:&writeError];
    
    if (writeSuccess) {
        return YES;
    } else {
        LOGGER_ERROR(@"CSV write error:%@", writeError);
        return NO;
    }
}

- (NSString *)generateContent {
    NSMutableString *content = [NSMutableString mutableStringWithByteOrderMark];
    [self appendReceiptsTable:content];
    [self appendDistancesTable:content];
    return content;
}

- (void)appendReceiptsTable:(NSMutableString *)content {
    ReportCSVTable *receiptsTable = [[ReportCSVTable alloc] initWithContent:content columns:[self receiptColumns]];
    [receiptsTable setIncludeHeaders:[WBPreferences includeCSVHeaders]];

    [receiptsTable appendTableWithRows:[self receipts]];
}

- (void)appendDistancesTable:(NSMutableString *)content {
    if (![WBPreferences printDistanceTable]) {
        return;
    }

    NSArray *distances = [self distances];
    if (distances.count == 0) {
        return;
    }

    [content appendString:@"\n\n"];

    ReportCSVTable *receiptsTable = [[ReportCSVTable alloc] initWithContent:content columns:[self distanceColumns]];
    [receiptsTable setIncludeHeaders:[WBPreferences includeCSVHeaders]];

    [receiptsTable appendTableWithRows:distances];
}

@end

File: SmartReceipts/Reports/Generators/TripCSVGenerator.swift
//
//  TripCSVGenerator.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 03/12/2017.
//  Copyright © 2017 Will Baumann. All rights reserved.
//

import Foundation

fileprivate let TABLE_SPACING = "\n \n"

class TripCSVGenerator: ReportCSVGenerator {
    override func generateTo(path: String) -> Bool {
        do { try generateContent().write(to: path.asFileURL) }
        catch {
            Logger.error("CSV write error: \(error.localizedDescription)")
           return false
        }
        return true
    }
    
    @objc private func generateContent() -> Data {
        let content = NSMutableString()
        
        appendReceiptsTable(content)
        content.append(TABLE_SPACING)
        
        appendDistancesTable(content)
        
        if WBPreferences.includeCategoricalSummation() && PurchaseService.hasValidSubscriptionValue {
            appendCategoricalSummationTable(content)
        }
        
        if WBPreferences.separatePaymantsByCategory() && PurchaseService.hasValidSubscriptionValue  {
            appendSeparatedByCategoryTable(content)
        }
        
        return (content as String).byteOrderMarked
    }
    
    private func appendReceiptsTable(_ content: NSMutableString) {
        let receiptTable = ReportCSVTable(content: content, columns: receiptColumns())!
        receiptTable.includeHeaders = true
        
        var recs = receipts()
        if WBPreferences.printDailyDistanceValues() {
            let dReceipts = DistancesToReceiptsConverter.convertDistances(distances()) as! [WBReceipt]
            recs.append(contentsOf: dReceipts)
            recs.sort(by: { $1.date.compare($0.date) == .orderedAscending })
        }
        receiptTable.append(withRows: recs)
    }
    
    private func appendCategoricalSummationTable(_ content: NSMutableString) {
        content.append(TABLE_SPACING)
        let receiptTable = ReportCSVTable(content: content, columns: categoryColumns())!
        receiptTable.includeHeaders = true
        let receipts = receiptsByCategories()
        receiptTable.append(withRows: Array(receipts.values))
    }
    
    private func appendSeparatedByCategoryTable(_ content: NSMutableString) {
        let categoryReceipts = receiptsByCategories()
        for (category, receipts) in categoryReceipts {
            let receiptTable = ReportCSVTable(content: content, columns: receiptColumns())!
            receiptTable.includeHeaders = true
            content.append(TABLE_SPACING)
            content.append(category + "\n")
            receiptTable.append(withRows: receipts)
        }
    }
    
    private func appendDistancesTable(_ content: NSMutableString) {
        if !WBPreferences.printDistanceTable() { return }
        let dists = distances()
        if dists.isEmpty { return }
        let receiptTable = ReportCSVTable(content: content, columns: distanceColumns())!
        receiptTable.includeHeaders = true
        receiptTable.append(withRows: dists)
    }
}

